Add in-house consensus making

terrycojones opened the pull request:
Here's a first pull request to add simple consensus generation from a BAM file.  I deleted the old code we had in `consensus.py` based on BLAST alignment results. I wrote that years ago and never finished it and it's now irrelevant.

TODO:
- Add more tests with different quality combinations.
- Make this usable from `make-consensus.py` (maybe I'll do that on the plane trip).

This is likely too simplistic in its treatment of BAM files. There are quite a few other things (based on SAM flags and the pysam processing) that could be a factor, though at this point I don't see why they would be and am not sure if they need to be considered. I think we'll have to try this and see how it goes and whether it breaks, etc.,

I had planned to make this a bit more general (with different `strategy` possibilities), but there may be no need, in which case the extra function in `dark/consensus.py` can just be merged into the main one.

This follows the treatment of quality scores used by Geneious, described at https://assets.geneious.com/manual/2020.1/static/GeneiousManualse43.html That may not be the best approach, but it is defensible (in a publication). Other options / nuances (and tests for them) could easi;y be added.

The testing framework is very basic, and may need to be made more sophisticated. But I was trying to get something simple working first. More might not be needed.

UdoGi commented on dark/consensus.py:
```diff
@@ -1,102 +1,88 @@
+        for read in column.pileups:
+            readCount += 1
+            base = read.alignment.query_sequence[read.query_position]
+            bases[base] += (
```
i'd find sth like this more readable/explicit: 
`bases[base] += 1 if ignoreQuality else read.alignment.query_qualities[read.query_position]`

terrycojones replied:
Yeah, I figured that line was going to get flagged. I originally wrote it pretty much as you suggest, then realized there was a cuter (= more obscure) way to do it...  I'll put it back. Thanks for reading closely :-)

UdoGi reviewed (commented):
if there are reads at the beginning/end of the reference sequence, they would be ignored right now? Couldn't it be that the sequenced virus has a few nucleotides more at one end?

terrycojones commented:
> if there are reads at the beginning/end of the reference sequence, they would be ignored right now? Couldn't it be that the sequenced virus has a few nucleotides more at one end?

Yes, good point. The original approach using pysam's pileup call didn't allow for this (or for detection of insertions). I've now switched to using the pysam `fetch` method, and added tests to handle reads that start before the reference or finish after it (or both).
